src/TableFunctions/TableFunctionFactory.cpp

```diff
@@ -16,16 +16,22 @@ namespace ErrorCodes
     extern const int LOGICAL_ERROR;
 }
 
+static const NameSet table_functions_allowed_in_readonly_mode
+{
+    "null", "view", "viewIfPermitted", "numbers", "numbers_mt", "generateRandom", "values", "cluster", "clusterAllReplicas"
+};
 
 void TableFunctionFactory::registerFunction(
     const std::string & name, TableFunctionCreator creator, Documentation doc, CaseSensitiveness case_sensitiveness)
```

Comment:
Maybe it would be better to replace `TableFunctionCreator` and `Documentation` with `TableFunctionData` similarly to AggregateFunctionFactory? 

This way we will force filling `allow_readonly` on the caller of registerFunction and get rid of `table_functions_allowed_in_readonly_mode`, which in my opinion is a good thing.

Reply:
agreed, that is better. addressed it.